# Activity: Explore descriptive statistics

## **Introduction**

Data professionals often use descriptive statistics to understand the data they are working with and provide collaborators with a summary of the relative location of values in the data, as well an information about its spread. 

For this activity, you are a member of an analytics team for the United States Environmental Protection Agency (EPA). You are assigned to analyze data on air quality with respect to carbon monoxide, a major air pollutant. The data includes information from more than 200 sites, identified by state, county, city, and local site names. You will use Python functions to gather statistics about air quality, then share insights with stakeholders.

## **Step 1: Imports** 


Import the relevant Python libraries `pandas` and `numpy`.

In [1]:
# Import relevant Python libraries.
import pandas as pd
import numpy as np

The dataset provided is in the form of a .csv file named `c4_epa_air_quality.csv`. It contains a susbet of data from the U.S. EPA. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.

### YOUR CODE HERE
epa_data = pd.read_csv("c4_epa_air_quality.csv", index_col = 0)

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about loading data in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to read in data from a .csv file and load it into a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `read_csv` function from the pandas `library`. The `index_col` parameter can be set to `0` to read in the first column as an index (and to avoid `"Unnamed: 0"` appearing as a column in the resulting DataFrame).

</details>

## **Step 2: Data exploration** 

To understand how the dataset is structured, display the first 10 rows of the data.

In [5]:
# Display first 10 rows of the data.

### YOUR CODE HERE
epa_data.head(10)

date_local    state_name   county_name      city_name   
0  2018-01-01       Arizona      Maricopa        Buckeye  \
1  2018-01-01          Ohio       Belmont      Shadyside   
2  2018-01-01       Wyoming         Teton  Not in a city   
3  2018-01-01  Pennsylvania  Philadelphia   Philadelphia   
4  2018-01-01          Iowa          Polk     Des Moines   
5  2018-01-01        Hawaii      Honolulu  Not in a city   
6  2018-01-01        Hawaii      Honolulu  Not in a city   
7  2018-01-01  Pennsylvania          Erie           Erie   
8  2018-01-01        Hawaii      Honolulu       Honolulu   
9  2018-01-01      Colorado       Larimer   Fort Collins   

                                     local_site_name   parameter_name   
0                                            BUCKEYE  Carbon monoxide  \
1                                          Shadyside  Carbon monoxide   
2  Yellowstone National Park - Old Faithful Snow ...  Carbon monoxide   
3                             North East Waste (NEW)  Carbon monoxide   
4                                          CARPENTER  Carbon monoxide   
5                                            Kapolei  Carbon monoxide   
6                                            Kapolei  Carbon monoxide   
7                                                NaN  Carbon monoxide   
8                                           Honolulu  Carbon monoxide   
9                      Fort Collins - CSU - S. Mason  Carbon monoxide   

    units_of_measure  arithmetic_mean  aqi  
0  Parts per million         0.473684    7  
1  Parts per million         0.263158    5  
2  Parts per million         0.111111    2  
3  Parts per million         0.300000    3  
4  Parts per million         0.215789    3  
5  Parts per million         0.994737   14  
6  Parts per million         0.200000    2  
7  Parts per million         0.200000    2  
8  Parts per million         0.400000    5  
9  Parts per million         0.300000    6

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about exploratory data analysis in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `head()` function from the `pandas` library.

</details>

**Question:** What does the `aqi` column represent?

[Write your response here. Double-click (or enter) to edit.]

**Question:** In what units are the aqi values expressed?

[Write your response here. Double-click (or enter) to edit.]

Now, get a table that contains some descriptive statistics about the data.

In [7]:
# Get descriptive stats.

### YOUR CODE HERE
epa_data.describe()

arithmetic_mean         aqi
count       260.000000  260.000000
mean          0.403169    6.757692
std           0.317902    7.061707
min           0.000000    0.000000
25%           0.200000    2.000000
50%           0.276315    5.000000
75%           0.516009    9.000000
max           1.921053   50.000000

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate a table of basic descriptive statistics about the numeric columns in a DataFrame.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `describe()` function from the `pandas` library.

</details>

**Question:** Based on the table of descriptive statistics, what do you notice about the count value for the `aqi` column?

[Write your response here. Double-click (or enter) to edit.]

**Question:** What do you notice about the 25th percentile for the `aqi` column?

This is an important measure for understanding where the aqi values lie. 

[Write your response here. Double-click (or enter) to edit.]

**Question:** What do you notice about the 75th percentile for the `aqi` column?

This is another important measure for understanding where the aqi values lie. 

[Write your response here. Double-click (or enter) to edit.]

## **Step 3: Statistical tests** 

Next, get some descriptive statistics about the states in the data.

In [10]:
# Get descriptive stats about the states in the data.

### YOUR CODE HERE
epa_data.groupby('state_name').agg({'arithmetic_mean': ['mean', 'median', 'std'], 'aqi': ['mean', 'median', 'std']})

arithmetic_mean                            aqi          
                                mean    median       std       mean median   
state_name                                                                   
Alabama                     0.200000  0.200000       NaN   2.000000    2.0  \
Alaska                      0.555264  0.555264  0.003722   8.500000    8.5   
Arizona                     0.671804  0.523684  0.519735  15.214286   10.0   
Arkansas                    0.300000  0.300000       NaN   3.000000    3.0   
California                  0.684871  0.641667  0.322950  12.121212   11.0   
Colorado                    0.330994  0.305263  0.071220   5.000000    5.0   
Connecticut                 0.200000  0.219444  0.069537   3.500000    4.0   
Delaware                    0.215789  0.215789       NaN   3.000000    3.0   
District Of Columbia        0.222222  0.222222  0.031427   2.500000    2.5   
Florida                     0.299123  0.310526  0.164393   5.500000    6.0   
Georgia                     0.286842  0.286842  0.122813   4.000000    4.0   
Hawaii                      0.531579  0.400000  0.413384   7.000000    5.0   
Idaho                       0.621053  0.621053  0.379605  10.000000   10.0   
Illinois                    0.254386  0.200000  0.147212   3.333333    2.0   
Indiana                     0.304210  0.247368  0.137820   3.800000    3.0   
Iowa                        0.171930  0.200000  0.062791   2.000000    2.0   
Kansas                      0.136842  0.136842       NaN   2.000000    2.0   
Kentucky                    0.181837  0.182353  0.018426   2.000000    2.0   
Louisiana                   0.000000  0.000000       NaN   0.000000    0.0   
Maine                       0.184210  0.184210  0.022330   2.500000    2.5   
Maryland                    0.195833  0.192106  0.015351   2.250000    2.0   
Massachusetts               0.225000  0.231579  0.033666   2.750000    3.0   
Michigan                    0.465400  0.415789  0.176400   8.111111    8.0   
Minnesota                   0.178947  0.200000  0.088227   2.142857    2.0   
Mississippi                 0.200000  0.200000       NaN   2.000000    2.0   
Missouri                    0.211111  0.200000  0.019245   3.000000    2.0   
Montana                     0.172222  0.172222  0.102137   2.000000    2.0   
Nebraska                    0.324416  0.324416  0.136666   7.000000    7.0   
Nevada                      0.988158  0.873684  0.603611  18.500000   13.0   
New Hampshire               0.200000  0.200000  0.000000   2.000000    2.0   
New Jersey                  0.302397  0.269444  0.153683   4.333333    4.0   
New Mexico                  0.239474  0.239474  0.011165   6.000000    6.0   
New York                    0.233684  0.210527  0.041234   2.500000    2.5   
North Carolina              0.191667  0.194445  0.010638   2.000000    2.0   
North Dakota                0.200000  0.200000       NaN   2.000000    2.0   
Ohio                        0.225379  0.238158  0.086912   3.333333    3.0   
Oklahoma                    0.171930  0.183334  0.102748   2.500000    2.0   
Oregon                      0.152632  0.152632  0.074432   2.000000    2.0   
Pennsylvania                0.239458  0.236764  0.122408   2.900000    3.0   
Puerto Rico                 0.336842  0.336842       NaN   5.000000    5.0   
Rhode Island                0.302941  0.302941  0.145581   4.000000    4.0   
South Carolina              0.100000  0.100000       NaN   1.000000    1.0   
South Dakota                0.242105  0.242105       NaN   5.000000    5.0   
Tennessee                   0.192105  0.192105  0.130257   2.000000    2.0   
Texas                       0.198947  0.200000  0.039682   2.700000    2.5   
Utah                        0.461306  0.363158  0.202742   9.000000   10.0   
Vermont                     0.227778  0.233333  0.125093   3.333333    3.0   
Virginia                    0.277632  0.284211  0.150614   4.000000    4.0   
Washington                  0.328948  0.328

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate basic descriptive statistics about a DataFrame or a column you are interested in.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

 Use the `describe()` function from the `pandas` library. Note that this function can be used:
- "on a DataFrame (to find descriptive statistics about the numeric columns)" 
- "directly on a column containing categorical data (to find pertinent descriptive statistics)"

</details>

**Question:** What do you notice while reviewing the descriptive statistics about the states in the data? 

Note: Sometimes you have to individually calculate statistics. To review to that approach, use the `numpy` library to calculate each of the main statistics in the preceding table for the `aqi` column.

[Write your response here. Double-click (or enter) to edit.]

## **Step 4. Results and evaluation**

Now, compute the mean value from the `aqi` column.

In [11]:
# Compute the mean value from the aqi column.

### YOUR CODE HERE
epa_data['aqi'].mean()

6.757692307692308

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the mean value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `mean()` function from the `numpy` library.

</details>

**Question:** What do you notice about the mean value from the `aqi` column?

This is an important measure, as it tells you what the average air quality is based on the data.

[Write your response here. Double-click (or enter) to edit.]

Next, compute the median value from the aqi column.

In [14]:
# Compute the median value from the aqi column.

### YOUR CODE HERE
epa_data['aqi'].median()

5.0

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the median value from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `median()` function from the `numpy` library.

</details>

**Question:** What do you notice about the median value from the `aqi` column?

This is an important measure for understanding the central location of the data.

[Write your response here. Double-click (or enter) to edit.]

Next, identify the minimum value from the `aqi` column.

In [15]:
# Identify the minimum value from the aqi column.

### YOUR CODE HERE
epa_data['aqi'].min()


0

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the minimum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `min()` function from the `numpy` library.

</details>

**Question:** What do you notice about the minimum value from the `aqi` column?

This is an important measure, as it tell you the best air quality observed in the data.

[Write your response here. Double-click (or enter) to edit.]

Now, identify the maximum value from the `aqi` column.

In [16]:
# Identify the maximum value from the aqi column.

### YOUR CODE HERE
epa_data['aqi'].max()

50

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the maximum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `max()` function from the `numpy` library.

</details>

**Question:** What do you notice about the maximum value from the `aqi` column?

This is an important measure, as it tells you which value in the data corresponds to the worst air quality observed in the data.

[Write your response here. Double-click (or enter) to edit.]

Now, compute the standard deviation for the `aqi` column.

By default, the `numpy` library uses 0 as the Delta Degrees of Freedom, while `pandas` library uses 1. To get the same value for standard deviation using either library, specify the `ddof` parameter to 1 when calculating standard deviation.

In [17]:
# Compute the standard deviation for the aqi column.

### YOUR CODE HERE
epa_data['aqi'].std()

7.061706678820724

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video section about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the standard deviation from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `std()` function from the `numpy` library. Make sure to specify the `ddof` parameter as 1. To read more about this function,  refer to its documentation in the references section of this lab.

</details>

**Question:** What do you notice about the standard deviation for the `aqi` column? 

This is an important measure of how spread out the aqi values are.

[Write your response here. Double-click (or enter) to edit.]

## **Considerations**


**What are some key takeaways that you learned during this lab?**

[Write your response here. Double-click (or enter) to edit.]

**How would you present your findings from this lab to others? Consider the following relevant points noted by AirNow.gov as you respond:**
- "AQI values at or below 100 are generally thought of as satisfactory. When AQI values are above 100, air quality is considered to be unhealthy—at first for certain sensitive groups of people, then for everyone as AQI values increase."
- "An AQI of 100 for carbon monoxide corresponds to a level of 9 parts per million."

[Write your response here. Double-click (or enter) to edit.]

**What summary would you provide to stakeholders? Use the same information provided previously from AirNow.gov as you respond.**

[Write your response here. Double-click (or enter) to edit.]

Overall, the standard deviation of the entire aqi dataset is about 7.06 (according to epa_data['aqi'].std()), which is relatively high considering the range of the AQI (0-50). 
This indicates that there's a significant amount of variability in the AQI across all states.

The data shows a wide range of standard deviations, meaning that the variability of air quality differs significantly from state to state. Some states have a more consistent air quality, while others have more variability.

**References**

[Air Quality Index - A Guide to Air Quality and Your Health](https://www.airnow.gov/sites/default/files/2018-04/aqi_brochure_02_14_0.pdf). (2014,February)

[Numpy.Std — NumPy v1.23 Manual](https://numpy.org/doc/stable/reference/generated/numpy.std.html)

US EPA, OAR. (2014, 8 July).[*Air Data: Air Quality Data Collected at Outdoor Monitors Across the US*](https://www.epa.gov/outdoor-air-quality-data). 